[jit] Fix dict type serialization

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32569 [jit] Fix dict type serialization**

If the dict's contained types cannot be inferred from its contents (for
example, `Dict[str, Tensor]` vs. `Dict[str, Optional[Tensor]]`), we must
explicitly annotate the type.

Also this removes some special handling that omits annotations on empty
containers that have the default type. It makes the code more complex
for not too much value, and was wrong for dicts anyway.

Differential Revision: [D19551016](https://our.internmc.facebook.com/intern/diff/D19551016)

suo committed:
[jit] Fix dict type serialization

If the dict's contained types cannot be inferred from its contents (for
example, `Dict[str, Tensor]` vs. `Dict[str, Optional[Tensor]]`), we must
explicitly annotate the type.

Also this removes some special handling that omits annotations on empty
containers that have the default type. It makes the code more complex
for not too much value, and was wrong for dicts anyway.

[ghstack-poisoned]

suo committed:
Update on "[jit] Fix dict type serialization"


If the dict's contained types cannot be inferred from its contents (for
example, `Dict[str, Tensor]` vs. `Dict[str, Optional[Tensor]]`), we must
explicitly annotate the type.

Also this removes some special handling that omits annotations on empty
containers that have the default type. It makes the code more complex
for not too much value, and was wrong for dicts anyway.

[ghstack-poisoned]

suo committed:
Update on "[jit] Fix dict type serialization"


If the dict's contained types cannot be inferred from its contents (for
example, `Dict[str, Tensor]` vs. `Dict[str, Optional[Tensor]]`), we must
explicitly annotate the type.

Also this removes some special handling that omits annotations on empty
containers that have the default type. It makes the code more complex
for not too much value, and was wrong for dicts anyway.

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -906,32 +906,40 @@ struct PythonPrintImpl {
-        if (node->inputs().size() == 0 && !is_default_type) {
+        // Empty dicts must be annotated with their type so the compiler knows
+        // what type is supposed to be inside them
+        if (node->inputs().size() == 0) {
```
This branch can be folded into the one below as another or clause. printDict knows know to print an empty Dict.

suo committed:
Update on "[jit] Fix dict type serialization"


If the dict's contained types cannot be inferred from its contents (for
example, `Dict[str, Tensor]` vs. `Dict[str, Optional[Tensor]]`), we must
explicitly annotate the type.

Also this removes some special handling that omits annotations on empty
containers that have the default type. It makes the code more complex
for not too much value, and was wrong for dicts anyway.

Differential Revision: [D19551016](https://our.internmc.facebook.com/intern/diff/D19551016)

[ghstack-poisoned]

suo committed:
Update on "[jit] Fix dict type serialization"


If the dict's contained types cannot be inferred from its contents (for
example, `Dict[str, Tensor]` vs. `Dict[str, Optional[Tensor]]`), we must
explicitly annotate the type.

Also this removes some special handling that omits annotations on empty
containers that have the default type. It makes the code more complex
for not too much value, and was wrong for dicts anyway.

Differential Revision: [D19551016](https://our.internmc.facebook.com/intern/diff/D19551016)

[ghstack-poisoned]

suo committed:
Update on "[jit] Fix dict type serialization"


If the dict's contained types cannot be inferred from its contents (for
example, `Dict[str, Tensor]` vs. `Dict[str, Optional[Tensor]]`), we must
explicitly annotate the type.

Also this removes some special handling that omits annotations on empty
containers that have the default type. It makes the code more complex
for not too much value, and was wrong for dicts anyway.

Differential Revision: [D19551016](https://our.internmc.facebook.com/intern/diff/D19551016)

[ghstack-poisoned]

suo committed:
Update on "[jit] Fix dict type serialization"


If the dict's contained types cannot be inferred from its contents (for
example, `Dict[str, Tensor]` vs. `Dict[str, Optional[Tensor]]`), we must
explicitly annotate the type.

Also this removes some special handling that omits annotations on empty
containers that have the default type. It makes the code more complex
for not too much value, and was wrong for dicts anyway.

Differential Revision: [D19551016](https://our.internmc.facebook.com/intern/diff/D19551016)

[ghstack-poisoned]
